cmd/hil-vpnd/mockprivops_test.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"crypto/rand"
+	"fmt"
+	"sync"
+)
+
+// A mock implementation of PrivOps, for testing. Use NewMockPrivOps to
+// create one of these; the zero value is not useful.
+type MockPrivOps struct {
+	lock sync.Mutex
+	vpns map[string]*vpnInfo
+}
+
+// Create a new MockPrivOps, with no existant vpns.
+func NewMockPrivOps() *MockPrivOps {
+	return &MockPrivOps{
+		vpns: make(map[string]*vpnInfo),
+	}
+}
+
+// Info about a vpn.
+type vpnInfo struct {
+	// The port number that openvpn would listen on
+	portNo uint16
+
+	// The vlan number for the network
+	vlanNo uint16
+
+	// The OpenVPN static key. For testing we just use a random
+	// string here.
+	key string
+
+	// Whether the vpn is up
+	running bool
+}
+
+// Generate a mock "key".
+func genMockKey() (string, error) {
+	var key [128 / 8]byte
+	_, err := rand.Read(key[:])
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%x", key), nil
+}
+
+// Get a vpnInfo for the named vpn, panicing if it doesn't exist.
+func (ops *MockPrivOps) mustGetVpn(name string) *vpnInfo {
+	vpn, ok := ops.vpns[name]
+	if !ok {
+		panic(fmt.Sprintf("Tried to operate on non-existent vpn %q", name))
+	}
+	return vpn
+}
+
+//// Implementations of the methods needed to implement the PrivOps interface.
+
+func (ops *MockPrivOps) CreateVPN(name string, vlanNo uint16, portNo uint16) (string, error) {
+	ops.startOp()
+	defer ops.endOp()
+	if _, ok := ops.vpns[name]; ok {
+		panic(fmt.Sprintf(
+			"Tried to create a vpn with the same name (%q) as an existing one.",
+			name,
+		))
+	}
+
+	key, err := genMockKey()
+	if err != nil {
+		return "", err
+	}
+
+	ops.vpns[name] = &vpnInfo{
+		portNo: portNo,
+		vlanNo: vlanNo,
+		key:    key,
+	}
+
+	return key, nil
+}
+
+func (ops *MockPrivOps) StartVPN(name string) error {
+	ops.startOp()
+	defer ops.endOp()
+	vpn := ops.mustGetVpn(name)
+	if vpn.running {
+		panic(fmt.Sprintf("Tried to start already-running vpn %q", name))
+	}
+	vpn.running = true
+	return nil
+}
+
+func (ops *MockPrivOps) StopVPN(name string) error {
+	ops.startOp()
+	defer ops.endOp()
+	vpn := ops.mustGetVpn(name)
+	if !vpn.running {
+		panic(fmt.Sprintf("Tried to stop vpn %q, which is not running.", name))
+	}
+	vpn.running = false
+	return nil
+}
+
+func (ops *MockPrivOps) DeleteVPN(name string) error {
+	ops.startOp()
+	defer ops.endOp()
+	if _, ok := ops.vpns[name]; !ok {
+		panic(fmt.Sprintf("Tried to delete non-existent vpn %q", name))
+	}
+	delete(ops.vpns, name)
```

Comment:
Should we check if a VPN is running before deleting it? Might come in handy for testing  if we want the check in real PrivOps